docs/extending/extending_client_side.md

```diff
@@ -0,0 +1,528 @@
+
+### Understanding the basics of Stimulus
+
+The [Stimulus documentation](https://stimulus.hotwired.dev/) is the best source on how to work with and understand Stimulus, here is a basic overview though.
+
+#### HTML first
+
+Consider the HTML structure of your components first, especially accessibility, Stimulus is built on top of HTML and will work best if you have a well-structured DOM.
+
+#### [Controllers](https://stimulus.hotwired.dev/reference/controllers)
+
+Controllers are [JavaScript classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) that extend the Stimulus base `Controller`. They get connected to an element with the `data-controller` attribute, one element can have multiple controllers.
+A Controller is registered with an identifier, a `kebab-case` string that allows the controller to be referenced in data attributes on your HTML, try to keep your Controllers small in scope within the DOM and in their functionality.
+Controllers have methods that define the main behaviour, plus a few built-in callbacks for `connect`, `disconnect`, and others specific to the values set up.
+
+#### [Targets](https://stimulus.hotwired.dev/reference/targets)
+
+Simple Controllers can use the `this.element` to get their controlled element, reach for targets if you need to have access to other DOM elements in your Controller code and avoid using DOM selectors within the JavaScript.
+
+Target elements are attached to the Controller instance with the `data-...-target` attribute, for example `<span data-my-controller-name-target="label"></span>`.
+
+#### [Values](https://stimulus.hotwired.dev/reference/values)
+
+Values are both the initial HTML-driven data for controllers and allow stateful/reactive behaviour. You must declare a value type and an optional default within your controller.
+Values are declared in the initial HTML on the controlled element with the `data-...-value` attribute, for example `<section data-controller="my-controller" data-my-controller-delay-value="300">...</section>`.
+
+#### [Actions](https://stimulus.hotwired.dev/reference/actions)
+
+A simple and powerful way to attach DOM event listeners to any element within your controlled element, letting you react to events such as `click` or `hover` and triggering a Controller method. No JavaScript is needed for this, it's all declarative in your HTML and Stimulus will clean up event listeners for you.
+
+The `data-action` attribute is used, for example `<button type="button" data-action="click->my-controller#open">Open</button>`, where `click` is the event, separated by an `->` and then the controller's identifier with a `#methodName`.
+
+### Adding a custom Stimulus controller
+
+Wagtail exposes a single client-side global (`window.Stimulus`), which is an instance of the core admin Stimulus application.
```

Comment:
How about we instead expose two globals?

```js
import * as Stimulus from "@hotwired/stimulus"

window.Stimulus = Stimulus;
window.stimulusApp = Stimulus.Application.start();
```

This goes against the examples in the Stimulus docs, but makes it much clearer which is a package and which is the instance, while avoiding the need for a custom `.use` which I think could easily be confused with an official API.

In the future this will also make it easier for us to support `import { Controller } from 'stimulus'` in ES modules.

Reply:
I'm confused by this. 

All the documentation I've read uses `window.Stimus` as the application instance and never `window.stimulusApp`. 

https://stimulus.hotwired.dev/handbook/installing#using-other-build-systems

Originally I had lowercase `window.application` for the instance (early PRs) and you advised this should follow the docs and be `window.Stimulus`.

I'm not sure what to do with exposing a base a `Controller`, I've thought about this a fair bit and tried the docs with a lot of variants of this and `Stimulus.use.Controller` felt the simplest. 

I agree we would love to see this as an es6 import but that's a while away and won't work with widjet media JS scripts. 

If you're pretty set on not having anything else on the application that may be a 'box' for us to put custom stuff in, maybe some other ideas;

- expose it as a standalone global
- expose it on the bare stimulus application instance

I'm just trying to be aware of the spectrum of users of this stuff but understand the goal of not confusing our Stimulus usage with the core docs.

Reply:
Have opted for your suggestion Thibaud, scrapping the `createController` util completely so we can just expose the module and then the application instance.

I would like `window.app` or `window.application` but will update the code to be `window.stimulusApp` as recommended. My reasoning is that `app` is shorter but I understand this may confuse.